daceml/onnx/op_implementations/img_op_implementations.py

```diff
@@ -232,129 +241,172 @@ def forward(node: ONNXOp, state: SDFGState,
+@op_implementation(op="BatchNormalization", name="pure")
+class PureBatchNormalization(ONNXForward):
+    @staticmethod
+    def forward_can_be_applied(node: ONNXOp, state: SDFGState,
+                               sdfg: SDFG) -> bool:
+        X = in_desc_with_name(node, state, sdfg, "X")
+        if len(X.shape) != 4:
+            return False
 
-        # hook up filter
-        read_W = new_state.add_read("W")
-        inner_filter_memlet = propagation.propagate_memlet(
-            new_state, filter_memlet, inner_me, False)
-        outer_filter_memlet = propagation.propagate_memlet(
-            new_state, inner_filter_memlet, outer_me, False)
-        new_state.add_edge(outer_me, None, inner_me, None, inner_filter_memlet)
-        new_state.add_edge(read_W, None, outer_me, None, outer_filter_memlet)
+        # only for training for now
+        if not {"out_mean", "out_var", "saved_mean", "saved_var"}.issubset(
+                node.out_connectors):
+            return False
+        if not {"scale", "B"}.issubset(node.in_connectors):
+            return False
 
-        # hook up X
-        read_X = new_state.add_read("X")
-        inner_image_memlet = propagation.propagate_memlet(
-            new_state, image_memlet, inner_me, False)
-        outer_image_memlet = propagation.propagate_memlet(
-            new_state, inner_image_memlet, outer_me, False)
-        new_state.add_edge(outer_me, None, inner_me, None, inner_image_memlet)
-        new_state.add_edge(read_X, None, outer_me, None, outer_image_memlet)
+        return True
 
-        # hook up outputs
-        output_memlet = dace.Memlet("Y[b, m, out_x, out_y]",
-                                    wcr="lambda x, y: x + y")
-        inner_output_memlet = propagation.propagate_memlet(
-            new_state, output_memlet, inner_me, False)
-        outer_output_memlet = propagation.propagate_memlet(
-            new_state, inner_output_memlet, outer_me, False)
-        new_state.add_edge(compute_tasklet, "output", inner_mx, None,
-                           output_memlet)
+    @staticmethod
+    def forward(node: ONNXOp, state: SDFGState,
+                sdfg: SDFG) -> typing.Union[nodes.Node, SDFG]:
 
-        write_Y = new_state.add_write("Y")
-        new_state.add_edge_pair(outer_mx, inner_mx, write_Y,
-                                inner_output_memlet, outer_output_memlet)
+        reduce_axes = list(
+            copy.deepcopy(in_desc_with_name(node, state, sdfg, "X").shape))
+        num_channels = reduce_axes.pop(1)
+
+        N = _prod(reduce_axes)
+        broadcast_shape = [num_channels, 1, 1]
+        dtype = in_desc_with_name(node, state, sdfg, "X").dtype
+        eps = node.epsilon
+        momentum = node.momentum
+        inv_momentum = 1 - node.momentum
+
+        def prog(X, scale, B, in_mean, in_var, Y, out_mean, out_var,
+                 saved_mean, saved_var):
+            saved_mean[:] = np.add.reduce(X, axis=[0, 2, 3]) / N
```

Comment:
these axes here should be computed outside (I think you can just use `reduce_axes`)